src/Microsoft.OData.Core/JsonLight/ODataJsonLightReader.cs

```diff
@@ -2256,6 +2259,117 @@ private void EndEntry()
             }
         }
 
+        /// <summary>
+        /// For response de-serialization, restore the null values for properties that are omitted
+        /// by the omit-values=nulls preference header.
+        /// </summary>
+        private void RestoreOmittedNullValues()
+        {
+            // Restore omitted null value only when processing response and when
+            // Preference header omit-values=nulls is specified.
+            if (ShouldRestoreNullValues())
+            {
+                IODataJsonLightReaderResourceState resourceState = this.CurrentResourceState;
+                IEdmStructuredType edmStructuredType = resourceState.ResourceType;
+
+                if (resourceState.SelectedProperties == SelectedPropertiesNode.Empty)
+                {
+                    return;
+                }
+                else
+                {
+                    if (resourceState.Resource != null && edmStructuredType != null)
+                    {
+                        ODataResourceBase resource = resourceState.Resource;
+
+                        IEnumerable<IEdmProperty> selectedProperties;
+
+                        if (resourceState.SelectedProperties == SelectedPropertiesNode.EntireSubtree)
+                        {
+                            selectedProperties = GetSelfAndBaseTypeProperties(edmStructuredType);
+                        }
+                        else
+                        {
+                           // Partial subtree.
+                           selectedProperties = resourceState.SelectedProperties.GetSelectedProperties(edmStructuredType);
+                        }
+
+                        // All dynamic properties: null value restoration.
+                        foreach (string name in resourceState.SelectedProperties.GetSelectedDynamicProperties(edmStructuredType))
+                        {
+                            RestoreNullODataProperty(name, resourceState);
+                        }
+
+                        foreach (IEdmProperty currentProperty in selectedProperties)
+                        {
+                            Debug.Assert(currentProperty.Type != null, "currentProperty.Type != null");
+
+                            // Skip declared properties that are not null-able types.
+                            // Skip navigation properties (no navigation link means link value is null)

```

Comment:
>(no navigation link means link value is null) [](start = 58, length = 45)

this is not true

Reply:
Presence or absence of navigation link doesn't have anything to do with whether or not the navigation property is null.

Navigation links are omitted by default in minimal metadata responses when they follow convention.  If you request full metadata, you will get back the link, even if the value of the navigation property is false.  For example:

  [https://services.odata.org/V4/TripPinService/People](https://services.odata.org/V4/TripPinService/People)('ronaldmundy')?$select=UserName,Photo&$format=application/json;odata.metadata=full

returns:
{
  "@odata.context": "[http://services.odata.org/V4/TripPinService/$metadata#People(UserName,Photo)/$entity](http://services.odata.org/V4/TripPinService/$metadata#People(UserName,Photo)/$entity)",
  "@odata.type": "#Microsoft.OData.SampleService.Models.TripPin.Person",
  "@odata.id": "[http://services.odata.org/V4/TripPinService/People(](http://services.odata.org/V4/TripPinService/People()'ronaldmundy')",
  "@odata.etag": "W/\"08D60D2CA6FBB30E\"",
  "@odata.editLink": "[http://services.odata.org/V4/TripPinService/People(](http://services.odata.org/V4/TripPinService/People()'ronaldmundy')",
  "UserName": "ronaldmundy",
  "Photo@odata.associationLink": "[http://services.odata.org/V4/TripPinService/People(](http://services.odata.org/V4/TripPinService/People()'ronaldmundy')/Photo/$ref",
  "Photo@odata.navigationLink": "[http://services.odata.org/V4/TripPinService/People(](http://services.odata.org/V4/TripPinService/People()'ronaldmundy')/Photo"
}

On the other hand, in minimal metadata, the links are omitted even if the value of the navigation property is non-null:

  [https://services.odata.org/V4/TripPinService/People](https://services.odata.org/V4/TripPinService/People)('russellwhyte')?$select=UserName,Photo

---
In reply to: [209789007](https://github.com/OData/odata.net/pull/1106#discussion_r209789007) [](ancestors = 209789007,198066211)

Reply:
Updated the inline comment. Thanks for the clarification!

---
In reply to: [213502295](https://github.com/OData/odata.net/pull/1106#discussion_r213502295) [](ancestors = 213502295,209789007,198066211)